Added GLM

Khalifa1997 opened the pull request:
in reponse to #5000 
The `train machine` function is currently empty, I will fill it up as soon as I get some feedback on a few things
In my head, I am using a `DescendUpdaterWithCorrection`/`DescendUpdater` inorder to have access to the function `update_variable` inside my class, yet. Which params should I pass to it, I know I could use `get_name()` to get which specific optimizer this and downcast it to this optimizer but I feel like there is a better way to go around it than this

Edit: I just saw that GLM.h has some extra doxygen comments, I will remove them on my next commit

Khalifa1997 committed:
Added GLM

gf712 commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,24 @@
+void GLM::init(){
+
+}
+GLM::GLM(DescendUpdaterWithCorrection* descend_updater, Family family, LinkFunction Link_fn, 
```
Could you use shared_ptr instead of raw pointer

gf712 replied:
Btw this should be the base class, ie DescendUpdater

vigsterkr replied:
second that

gf712 commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,104 @@
+{
+enum Family
+{
+	normal,
```
We usually write the enum values in caps

gf712 commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,104 @@
+		 *
+		 * @return whether training was successful
+		 */
+		virtual bool train_machine(Features* data=NULL) {return true;};
```
Same here, shouldn’t this be shared_ptr? I think this is actually not overriding the base class train_machine but actually creating a new function that is virtual

vigsterkr replied:
afaik this is protected in Machine. i mean the version with shared_ptr<Features>

gf712 replied:
Yea that is the issue that #5001 would flag at compile time. Should use override and then the compiler complains if there is no virtual function with the same signature in the base class.

gf712 commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,24 @@
+using namespace shogun;
+
+GLM::GLM(): LinearMachine(){
+    init();
```
The class members would be registered here with watch_method/SG_ADD

gf712 replied:
Sorry, not here, below in ::init()

gf712 commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,104 @@
+		/** Default constructor */
+		GLM();
+
+		/** default constructor
```
This isn’t a default constructor though

gf712 reviewed (commented):
I think it’s a good start, but need to see how it will be implemented in train_machine!

vigsterkr commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,104 @@
+
+		/** @return object name */
+		virtual const char* get_name() const { return "GLM"; }
+
```
no need for this many new lines :)

vigsterkr commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,104 @@
+
+	protected:
+
+		DescendUpdaterWithCorrection* m_descend_updater;
```
shared_ptr plz

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,24 @@
+}
+GLM::GLM(DescendUpdaterWithCorrection* descend_updater, Family family, LinkFunction Link_fn, 
+	float64_t alpha, float64_t lambda): LinearMachine(){
+        m_alpha= alpha;
```
nitpicking: plz be consistent with equation either `<variable>= value` or `<variable> = value`

vigsterkr commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,104 @@
+	public:
+		/** problem type */
+		MACHINE_PROBLEM_TYPE(PT_REGRESSION);
+
```
clang-format?

Khalifa1997 committed:
Requested changes

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -7,18 +7,32 @@
-    }
\ No newline at end of file
+GLM::GLM(
+    std::shared_ptr<DescendUpdater> descend_updater, Family family,
```
`const std::shared_ptr<DescendUpdater>& descend_updater`

Khalifa1997 replied:
funny it built the first time, changed it tho

vigsterkr commented on src/shogun/regression/GLM.h:
```diff
@@ -17,88 +17,96 @@
+		 * @param alpha alpha
+		 * @param lambda lambda
+		 */
+		GLM(std::shared_ptr<DescendUpdater>,
```
`const std::shared_ptr<DescendUpdater>& descend_updater`

Khalifa1997 committed:
added const std::shared_ptr<DescendUpdater>&

karlnapf commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,112 @@
+		 * https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/family
+		 * @param Family the family check
+		 * https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/family
+		 * @param alpha alpha
```
what are alpha and lambda? could you also describe them in the docs if you put docs? (otherwise they are pointless)

karlnapf commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,112 @@
+		 * @param alpha alpha
+		 * @param lambda lambda
+		 */
+		GLM(const std::shared_ptr<DescendUpdater>& descend_updater,
```
could you remove those ctors? The new API design uses the default ctor with subsequent `put` calls

vigsterkr replied:
in SWIG... i would still have this for c++

karlnapf replied:
fair enough. in that case, please make sure the class works with the default ctor. We had/have many bugs where a class segfaults in the train method as the default ctor didnt initialize things and the train method didnt check that. Don't rely that a user used this ctor, write separate tests for both

karlnapf replied:
and also please make sure that the default parameters in the ctor match the default parameters when using the default one (another source of confusion). I would remove the default values in the ctor to be on the safe side actually, and instead you can initialize the values in the `::init` method

gf712 replied:
You can even have the values inline with the class member declaration

karlnapf commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,112 @@
+		 * @param lambda lambda
+		 */
+
+		/** default destructor */
```
no need for such comments

karlnapf commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,112 @@
+		 * @param labs labels
+		 * @param learn_rate Learning rate
+		 * @param Linkfn the link function check
+		 * https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/family
```
such references would go into the class docs, not the ctor docs

karlnapf commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,112 @@
+		 * @param lambda lambda
+		 */
+		GLM(const std::shared_ptr<DescendUpdater>& descend_updater,
+		    Family family = POISSON_DISTRIBUTION, LinkFunction Link_fn = LOG,
```
Link_fn -> link_fn

karlnapf commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,112 @@
+		float64_t m_alpha;
+		float64_t m_lambda;
+		Family m_family;
+		LinkFunction m_linkfn;
```
pls use consistent variable names, i.e. "m_link_fn"

karlnapf commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,112 @@
+		NORMAL_DISTRIBUTION,
+		EXPONENTIAL_DISTRIBUTION,
+		GAMMA_DISTRIBUTION,
+		BINOMIAL_DISTRIBUTION,
```
As I said before, I am not very sure of representing these things as enums, this will lead to spaghetti code with that many. BUT you can leave it for now and just focus on the poisson regression. Once that is done we can think about this again

geektoni replied:
We could make the `GLM` class accept a `GLMDistribution` object instead. It will contain the log-likelihood and gradients of the given distribution (e.g., `GLMDistributionPoisson`). Then, the `GLM` class will only call the methods of the `GLMDistribution` to train itself.  

However, this is obviously out of the scope of this PR.

karlnapf replied:
that, or have locally defined helper classes that are then instantiated based on an enum.
But the code for single cases (eg posison) should be in a single place (likelihood contribution, gradients, etc)

karlnapf commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,112 @@
+
+namespace shogun
+{
+	enum Family
```
if you put this enum into the shogun namespace, it should have a better name

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,38 @@
+
+void GLM::init()
+{
+	SG_ADD(&m_alpha, "alpha", "alpha parameter", ParameterProperties::HYPER);
```
please avoid empty comments/docs/etc as "alpha parameter" for a varialbe called "alpha". If you put a description, why not say what it is or what it does?

karlnapf commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,112 @@
+ */
+/*
+References:
+https://www.rdocumentation.org/packages/stats/versions/3.6.2/topics/family
```
I'd put this into the class docs.
But actually, don't just blindly put references, we dont really need reference to other libs in here. Rather a reference to wikipedia describing GLM (or Poisson regression) in general, in a conceptual/mathy sense

karlnapf reviewed (commented):
This is good progress from the first version! But they key thing now will be to write a train method. I suggest you add a log_likelihood method first, and then one that computes the gradient of it. And then finally you can call the descent updater in the train method. See wikipedia for the expressions. As said, I would start to do this for Poisson regression first, and then once it works we can modularise it

karlnapf commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,112 @@
+#include <shogun/lib/config.h>
+
+#include <shogun/features/Features.h>
+#include <shogun/machine/FeatureDispatchCRTP.h>
```
good point, would be good if this class worked for multiple feature types. But you are not using this atm I think?

Khalifa1997 replied:
I haven't explored all the feature types aren't all features the same? (aka. just numbers at the end of the day) please do enlighten me :D

karlnapf replied:
read the code :)

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,38 @@
+	m_lambda = lambda;
+	m_linkfn = Link_fn;
+	m_descend_updater = descend_updater;
+	m_family = family;
```
you need to call the init method here  as well

vigsterkr replied:
i would rather call the default GLM above... that way no need to call init.

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,38 @@
+GLM::GLM(
+    const std::shared_ptr<DescendUpdater>& descend_updater, Family family,
+    LinkFunction Link_fn, float64_t alpha, float64_t lambda)
+    : LinearMachine()
```
```suggestion
    : GLM()
```

vigsterkr replied:
this way init will be called before setting the new values

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,38 @@
+	SG_ADD(&m_family, "family", "family used", ParameterProperties::SETTING);
+	SG_ADD(
+	    &m_linkfn, "linkfn", "Link function used",
+	    ParameterProperties::SETTING);
```
you need to initialize all values to defaults here. This brings up an interesting question. What is the default descend updater? I think it would be good to have one set so that users are not forced to pass one (tedious).

Khalifa1997 replied:
I think since they are just linear models at the end of the day, there isn't really alot of parameters to learn like in Neural Networks, SGD would work best here which I believe is `GradientDescendUpdater`

karlnapf replied:
The loss is convex (for Poisson regression with log link function), so a second order method like Newton will be better. But we can change that later, best is to start with something simple, then work it up from there.

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,38 @@
+	    ParameterProperties::SETTING);
+}
+GLM::GLM(
+    const std::shared_ptr<DescendUpdater>& descend_updater, Family family,
```
Another comment as I saw a concversation in irc re swig and the fact that the descend updater is not exposed yet. I would not worry about swig and the interfaces for now. Let's focus on the c++ code for now, possibly with unit tests.
The next step here (after incorporating the feedback you got) is to implement the gradient of the log-likelihood, and the next thing then is passing that to a descent updater. For the latter, you can start with plain old gradient descent, no need to be more fancy (for now).

Khalifa1997 committed:
Additional Changes

Khalifa1997 commented:
I have updated the class I will try to list all the changes I have made
1. Changed the ctor structure such that the default values are inline in the class variable declarations so the default ctor problem no longer exists
2. Removed pointless comments and changed variable naming + added some description to values like alpha/lambda
3. I tried implementing log_likelihood and I know how, based on [this](http://glm-tools.github.io/pyglmnet/tutorial.html#poisson-log-likelihood) the only problem is I was using `SGVector` and `SGMatrix` but since I am going to have to use things like the dot operator maybe using `Features` class would be the better idea here - and label class for the label- but Idk which type of `Features` should I be using, `DotFeatures`?
4. Have yet to decide a default `DescendUpdater`

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,55 @@
+SGVector<float64_t>
+log_likelihood(SGVector<float64_t> features, float64_t label)
+{
+	ASSERT(features.size() > 0)
```
You would write code that implements the log likelihood for all data at once

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,55 @@
+	// Assume weights is the same as the feature vector
+	SGVector<float64_t> weights = SGMatrix(features.clone());
+	// Assume b0 to be the element in the beginning of the feature vector
+	float64_t b0 = features.get_element(0);
```
this code doesnt do anything yet

karlnapf commented:
I just read through the code, and many of the things we commented on are not addressed yet.

RE your question about the features, read some other classes, e.g. ridge regression, that will make it clear

Khalifa1997 committed:
Updated log_likelihood

Khalifa1997 commented:
I have updated the log_likelihood function such that it takes dense features, and computes the likelihood, as you could tell tho the naming of variables is abit off, so I am open to suggestions in that area. As for the gradient calculation I am using [this](http://glm-tools.github.io/pyglmnet/tutorial.html#gradient-descent) implementation, however I am not sure if shogun has a sigmoid function or would I have to make a small function for it?

Khalifa1997 commented on src/shogun/regression/GLM.cpp:
```diff
@@ -7,18 +7,32 @@
+{
+	SG_ADD(&m_alpha, "alpha", "alpha parameter", ParameterProperties::HYPER);
+	SG_ADD(&m_lambda, "lambda", "lambda parameter", ParameterProperties::HYPER);
+	SG_ADD(
+	    (std::shared_ptr<SGObject>*)&m_descend_updater, "descend_updater",
+	    "Descend Updater used for updating weights",
+	    ParameterProperties::SETTING);
```
as for those lines, I think that's what causing the build to fail

Khalifa1997 replied:
any thoughts? @karlnapf

karlnapf replied:
Not sure what you mean

Khalifa1997 replied:
I get an error when I test as you could tell in the CI, in the SGObject test, so I have a feeling this is what causing it. what do you think?

vigsterkr replied:
when you report such thing, could you please make sure that you copy here the link to the CI line where you think the error is....

gf712 replied:
The issue is probably that m_descend_updater (`DescendUpdater`) is never initialised, so you are serialising a nullptr, and I think that might be causing issues... @vigsterkr ?

gf712 replied:
Hence https://dev.azure.com/shogunml/shogun/_build/results?buildId=3277&view=logs&j=c857963e-3b4d-5192-8776-ae81779de3af&t=542541c1-2650-5e44-f82f-a76d8c4f5255&l=216

vigsterkr replied:
mmm but this should catch that:
https://github.com/shogun-toolbox/shogun/blob/develop/src/shogun/io/serialization/JsonSerializer.cpp#L178

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,82 @@
+	init();
+}
+SGVector<float64_t> log_likelihood(
+    const std::shared_ptr<DenseFeatures<float64_t>> features,
+    std::shared_ptr<Labels> label)
```
```suggestion
    const std::shared_ptr<DenseFeatures<float64_t>>& features,
    const std::shared_ptr<Labels>& label)
```

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,82 @@
+    const std::shared_ptr<DenseFeatures<float64_t>> features,
+    std::shared_ptr<Labels> label)
+{
+	int32_t vector_count = features->get_num_vectors();
+	int32_t feature_count = features->get_num_features();
```
```suggestion
	auto vector_count = features->get_num_vectors();
	auto feature_count = features->get_num_features();
```

vigsterkr replied:
in general try to use for these kind of statements auto

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,82 @@
+	SGVector<float64_t> lambda(vector_count);
+	for (int32_t i = 0; i < vector_count; i++)
+	{
+		SGVector<float64_t> feature_vector = features->get_feature_vector(i);
```
auto

karlnapf replied:
@Khalifa1997 why do you resolve conversations when you haven't addressed the comment by @vigsterkr ?
This just results in us giving redundant feedback -> wasting time

Khalifa1997 replied:
he had mentionned auto to be used in the for loop and I think I had address them all in my code +he mentionned the ctor thing when it took default values. I fixed this by giving default values inline with member attributes. but if I had missed any comment, then yea I appologise and ofc u can point it out :)

karlnapf replied:
You can always use auto if possible. The only case where I would write the type is if it would be unclear/confusing otherwise :)

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,82 @@
+	}
+	SGVector<float64_t> likelihood(vector_count);
+
+	for (int32_t i = 0; i < vector_count; i++)
```
this whole thing does the following:
likelihood = y*log(lambda) - lambda
this is definitely vectorizable... please use linalg framework to do this. instead of doing this vanilla code

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,82 @@
+		likelihood.set_element(
+		    y_i * log(lambda.get_element(i)) - lambda.get_element(i), i);
+	}
+	SGVector<float64_t> likelihood_clone = likelihood.clone();
```
auto

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,82 @@
+		// Assume beta0 is the same as the first element in the feature vector
+		float64_t b0 = feature_vector.get_element(0);
+		SGVector<float64_t> result(feature_count);
+		SGVector<float64_t>::vector_multiply(
```
linalg

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,82 @@
+		SGVector<float64_t> result(feature_count);
+		SGVector<float64_t>::vector_multiply(
+		    result, feature_vector, beta, feature_vector.vlen);
+		float64_t res = SGVector<float64_t>::sum(result);
```
linalg

vigsterkr reviewed (changes requested):
initial problems...

Khalifa1997 committed:
Vectorized code

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,80 @@
+		SGVector<float64_t> feature_vector = features->get_feature_vector(i);
+		// Assume beta is the same as the feature vector
+		SGVector<float64_t> beta = feature_vector.clone();
+		// Assume beta0 is the same as the first element in the feature vector
```
we don't do that, this is a trick to make the math look easier, but we will treat bias and weights separately please

Khalifa1997 replied:
you mean use `bias` and `m_w` for weights?

lgoetz replied:
yes, for example; the point is, you should not assume the bias is the first element of the feature vector

Khalifa1997 replied:
@lgoetz yes I have changed that in my last commit thanks :)

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,80 @@
+	ASSERT(vector_count > 0 && label->get_num_labels() == vector_count)
+	// Array of Lambdas
+	SGVector<float64_t> lambda(vector_count);
+	for (auto i = 0; i < vector_count; i++)
```
you don't need to loop here if using dense features. You can do a matrix vector product

Khalifa1997 replied:
there is no `log` operation in `linalg` thus I don't think it can be done without a loop

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,80 @@
+	SGVector<float64_t> log_lambda(vector_count);
+
+	for (auto i = 0; i < vector_count; i++)
+		log_lambda.set_element(log(lambda.get_element(i)), i);
```
don't we have element wise log in linalg? if not we should have it

Khalifa1997 replied:
I don't think there is. I had checked in the documentation and didn't find anything for it

lgoetz replied:
see src/shogun/distributions/LinearHMM.cpp for example

karlnapf replied:
If it is not in linalg, add it there. We have something for elementwise ops in there iirc.
I don't think LinearHMM is a good place to look .. It is way too old school ;)

Khalifa1997 replied:
@karlnapf do I start another PR and add it?

lgoetz replied:
> If it is not in linalg, add it there. We have something for elementwise ops in there iirc.
> I don't think LinearHMM is a good place to look .. It is way too old school ;)

sorry ;)

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,80 @@
+	    linalg::element_prod(labels, log_lambda), lambda, 1.0, -1.0);
+	SGVector<float64_t> likelihood_clone = likelihood.clone();
+	for (auto i = 0; i < vector_count; i++)
+		likelihood.set_element(
```
same here, please do with this matrix algebra and elementwise operations

Khalifa1997 replied:
I have changed this part to return `linalg::sum(likelihood)` so should be done

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,80 @@
+	auto vector_count = features->get_num_vectors();
+	auto feature_count = features->get_num_features();
+	ASSERT(vector_count > 0 && label->get_num_labels() == vector_count)
+	// Array of Lambdas
```
remove such comments

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,80 @@
+}
+void GLM::init()
+{
+	SG_ADD(
```
this is for the elastic net stuff, which we don't need here for now. Just the L2 reguariser is enough

Khalifa1997 replied:
so just removing alpha and lambda and adding tau as in `LinearRidgeRegression` ?

lgoetz replied:
Yes (or use lambda instead of tau). And please add "L2" to the description to clarify it's a ridge penalty

Khalifa1997 replied:
Done, I think I will use tau for this method to keep it consistent with the other models such as LinearRidgeRegression + I am using the naming lambda for the likelihood array

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,80 @@
+	    &m_alpha, "alpha", "Weighting parameter between L1 and L2 Penalty",
+	    ParameterProperties::HYPER);
+	SG_ADD(
+	    &m_lambda, "lambda", "Regularization parameter lambda",
```
Add "L2".

karlnapf reviewed (commented):
well done on adding the log likelihood. Though you should vectorize it.

Next step is to implement the gradient of that

Khalifa1997 committed:
Removed Alpha, added tau and More vectorization

Khalifa1997 committed:
Removed Lambda + added L2 to the description

Only using tau to stay consistent with the naming convention of other regression models

Khalifa1997 committed:
Added likelihood derivative for beta0

Khalifa1997 commented:
Now I am going to briefly sum up what I have added here:
the expected result of the log_likelihood_derivative should be of count `len(beta) + 1 //for the beta0 element`  so I have made a result array for those things.
Now I am facing two problems atm
1. there is no log function in linalg
2. there is no elementwise division in linalg

now on line 70 in the .cpp file  it should of been `y*s / q` but since there is no linalg I would have to do it without vectorization which would make the code much much more ugly.

So I would love to have some feedback on what to do now thanks!

N.B: implementing grad_beta should be very straight forward once the issues I have mentionned are solved

Khalifa1997 committed:
Converted likelihood to a number +  changed likelihood derivative

I misunderstood likelihood to be an array but it's supposed to be a number (per B0,Beta) thus cleaned this up + changed stuff to likelihood_derivative so that it runs

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,100 @@
+	float64_t beta0 = LinearMachine::get_bias();
+	for (auto i = 0; i < vector_count; i++)
+	{
+		SGVector<float64_t> feature_vector = features->get_feature_vector(i);
```
auto

Khalifa1997 replied:
Done.

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,100 @@
+	for (auto i = 0; i < vector_count; i++)
+	{
+		SGVector<float64_t> feature_vector = features->get_feature_vector(i);
+		float64_t res = linalg::dot(feature_vector, beta);
```
dont do a loop here but a matrix vector product

Khalifa1997 replied:
Done.

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,100 @@
+    const std::shared_ptr<DenseFeatures<float64_t>>& features,
+    const std::shared_ptr<Labels>& label)
+{
+	auto vector_count = features->get_num_vectors();
```
it would be good if you could modularise this.
The log likelihood consists of multiple parts, activation, data terms, etc. You could put each of those into a helper method to make it more readable

Khalifa1997 replied:
Okay will do that in my next commit thanks:)

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,100 @@
+	SGVector<float64_t> q(vector_count);
+	for (auto i = 0; i < vector_count; i++)
+	{
+		q.set_element(log(1 + std::exp(z.get_element(0, i))), i);
```
q[i] z[i]

gf712 replied:
And use linalg?

Khalifa1997 replied:
@gf712  there is no linalg for log :S,
and by your comment @karlnapf do you mean this is how I should get the elements?

karlnapf reviewed (commented):
Cool, great that you have the gradient in place now

You should slowly start thinking about a unit test where you compare likelihood / gradient values with a reference implementation on some fixed simple data

Also made some minor comments

gf712 commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,100 @@
+
+	SGVector<float64_t> lambda(vector_count);
+	SGVector<float64_t> beta = LinearMachine::get_w();
+	float64_t beta0 = LinearMachine::get_bias();
```
Do we override get_w/get_bias? As in is there a particular reason why you specify LinearMachine::?

Khalifa1997 replied:
For some reason, I get get_w not declared in this scope without me doing so

Edit: now it works after I fixed some of my includes.. ig :D

vigsterkr replied:
i'm not so sure how includes changed things...

Khalifa1997 committed:
Used auto + Vectorization + Access SGVec/SGMat using [ ]

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,106 @@
+	auto feature_count = features->get_num_features();
+	ASSERT(vector_count > 0 && label->get_num_labels() == vector_count)
+	SGVector<float64_t> result(vector_count + 1);
+	SGVector<float64_t> beta = LinearMachine::get_w();
```
auto....

vigsterkr replied:
get_w() should just work.

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,106 @@
+	ASSERT(vector_count > 0 && label->get_num_labels() == vector_count)
+	SGVector<float64_t> result(vector_count + 1);
+	SGVector<float64_t> beta = LinearMachine::get_w();
+	float64_t beta0 = LinearMachine::get_bias();
```
auto

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,106 @@
+	SGVector<float64_t> result(vector_count + 1);
+	SGVector<float64_t> beta = LinearMachine::get_w();
+	float64_t beta0 = LinearMachine::get_bias();
+	SGMatrix<float64_t> z = linalg::matrix_prod(
```
auto

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,106 @@
+	{
+		q[i] = log(1 + exponent[i]);
+	}
+	float64_t beta0_grad =
```
auto

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,106 @@
+GLM::GLM(
+    const std::shared_ptr<DescendUpdater>& descend_updater,
+    DistributionFamily family, LinkFunction link_fn, float64_t tau)
+    : LinearMachine()
```
call `GLM()` instead of LinearMachine()... this way you dont need the redundancy to call init()....

vigsterkr commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,106 @@
+	m_link_fn = link_fn;
+	m_descend_updater = descend_updater;
+	m_family = family;
+	init();
```
remove it once you've fixed the ctor i've mentioned above.

vigsterkr commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,91 @@
+		GLM(const std::shared_ptr<DescendUpdater>& descend_updater,
+		    DistributionFamily family, LinkFunction link_fn, float64_t tau);
+
+		virtual ~GLM(){};
```
i have explicitly stated that these should be `override` instead of virtual. you have marked this resolved, but it has not been changed at all....

vigsterkr commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,91 @@
+		 * @param data training data
+		 * @return whether training was successful
+		 */
+		virtual bool train_machine(std::shared_ptr<Features> data = NULL)
```
it's still protected... and as said before if you would be using `override` then you would get an error actually

vigsterkr commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,91 @@
+		};
+
+		/** @return object name */
+		virtual const char* get_name() const
```
`override`...

vigsterkr commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,91 @@
+		MACHINE_PROBLEM_TYPE(PT_REGRESSION);
+
+		GLM();
+		float64_t log_likelihood(
```
this is nitpicking but still: ctors should be first and then class methods... although i'm not sure whether these really need to be public?

Khalifa1997 committed:
Addressed comments

Added Autos wherever possible + changes to ctor + override

vigsterkr commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,96 @@
+		    const std::shared_ptr<DenseFeatures<float64_t>>& features,
+		    const std::shared_ptr<Labels>& label);
+
+		virtual bool
```
you can just drop `virtual`

vigsterkr commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,96 @@
+		};
+
+		/** @return object name */
+		virtual const char* get_name() const override
```
you can drop virtual here as well.... but get_name is public.

vigsterkr commented on src/shogun/regression/GLM.h:
```diff
@@ -0,0 +1,96 @@
+		GLM(const std::shared_ptr<DescendUpdater>& descend_updater,
+		    DistributionFamily family, LinkFunction link_fn, float64_t tau);
+
+		virtual ~GLM() override{};
```
you can drop virtual...

Khalifa1997 committed:
Removed virtual and added get_name to public

karlnapf commented on src/shogun/regression/GLM.cpp:
```diff
@@ -0,0 +1,105 @@
+	auto exponent = linalg::exponent(res);
+	for (auto i = 0; i < vector_count; i++)
+	{
+		lambda[i] = log(1 + exponent[i]);
```
std::

gf712 commented:
Implemented in #5005
